File: ModelManager/trainers/__task_event_trainer.py
# author: Khanh.Quang 
# institute: Hanoi University of Science and Technology
# file name: __task_event_trainer.py
# project name: ModelManager
# date: 01/12/2020
from pyspark.ml import Estimator
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.clustering import GaussianMixture
from pyspark.ml.feature import Imputer
from pyspark.ml.pipeline import Pipeline

from loaders import TaskEventDataLoader, DataLoader
from trainers.__trainer import Trainer


class TaskEventTrainer(Trainer):

    def _get_model_path(self) -> str:
        return "hdfs://localhost:9000/data/models/task-event"

    def _get_preprocessor(self):
        cols = [col for col in self._get_cols()]
        imputer = Imputer(inputCols=cols, outputCols=cols)
        vector_assembler =  VectorAssembler(inputCols=cols,
                                            outputCol="features").setHandleInvalid("keep")
        return Pipeline(stages=[imputer, vector_assembler])

    def _get_data_loader(self) -> DataLoader:
        return TaskEventDataLoader()

    def _get_data_path(self) -> str:
        return "hdfs://localhost:9000/data/task-event"

    def _create_estimator(self) -> Estimator:
        model = GaussianMixture().setK(3).setSeed(1)
        return model

    def _get_cols(self) -> list:
        return ['_c10', '_c11', '_c12']

File: ModelTrainer/src/main/scala/com/hust/modeltrainer/trainers/TaskEventTrainer.scala
package com.hust.modeltrainer.trainers

import com.hust.modeltrainer.iostream.{SparkDumper, SparkLoader}
import org.apache.spark.ml.clustering.{GaussianMixture, GaussianMixtureModel}
import org.apache.spark.ml.{Estimator, Pipeline, Transformer}
import org.apache.spark.ml.feature.{Imputer, VectorAssembler}
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions.col

class TaskEventTrainer(var modelPath: String) extends Trainer {

  override def getDataPath: String =
    "hdfs://hadoop-namenode:8020/data/task-event/"

  override def getModelPath: String =
    modelPath

  override def train: Unit = {
    val inputDF = SparkLoader.csv(getDataPath).select("_c10", "_c11", "_c12")
    val inputDFCasted = inputDF.select(
      inputDF.columns.map((c: String) => col(c).cast("float")): _*
    ).limit(10000)
    val transformer = createTransformer(inputDFCasted)
    val predictor = createPredictedModel

    val pipeline = new Pipeline().setStages(Array(transformer, predictor))

    val model = pipeline.fit(inputDFCasted)

    model.transform(inputDFCasted).show(10)

    SparkDumper.dump(model, getModelPath)
  }

  def createPredictedModel: Estimator[GaussianMixtureModel] = {
    new GaussianMixture().setK(3)
  }

  def createTransformer(df: DataFrame): Pipeline = {
    val imputedComlumns = getImputedColumns(df)

    val imputer = new Imputer()
      .setInputCols(df.columns)
      .setOutputCols(imputedComlumns)
      .setStrategy("mean")

    val vectorAssembler = new VectorAssembler()
      .setInputCols(imputedComlumns)
      .setOutputCol("features")

    new Pipeline().setStages(Array(imputer, vectorAssembler))
  }

  private def getImputedColumns(df: DataFrame): Array[String] =
    df.columns.map((c: String) => c + "Imputed")
}
